File: src/main/java/com/truveta/opentoken/tokens/NotInValidator.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens;

import java.util.Arrays;
import javax.validation.constraints.NotNull;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

/**
 * A Validator that asserts that the attribute values is
 * <b>NOT IN</b> the list of invalid values.
 */
@AllArgsConstructor
@Getter
@Setter
public final class NotInValidator implements AttributeValidator {

    @NotNull
    private String attributeName;
    @NotNull
    private String[] invalidValues;

    /**
     * Validates that the attribute value is not in the list of invalid values.
     */
    @Override
    public boolean eval(String name, String value) {
        return (!name.equals(attributeName)) || !Arrays.asList(invalidValues).contains(value);
    }

}

File: lib/nodejs/opentoken/src/attributes/validation/NotInValidator.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { AttributeValidator } from './AttributeValidator';

/**
 * A validator that checks if a value is NOT in a given set of invalid values.
 */
export class NotInValidator implements AttributeValidator {
  private invalidValues: Set<string>;

  /**
   * Constructs a new NotInValidator.
   *
   * @param invalidValues - Set of values that are considered invalid
   */
  constructor(invalidValues: Set<string> | string[]) {
    this.invalidValues = invalidValues instanceof Set ? invalidValues : new Set(invalidValues);
  }

  /**
   * Validates that the given value is NOT in the set of invalid values.
   *
   * @param value - The value to validate.
   * @returns true if the value is NOT in the invalid values set; false otherwise.
   */
  validate(value: string): boolean {
    if (!value) {
      return false;
    }
    return !this.invalidValues.has(value);
  }
}
